### DISCUSSION TITLE
Unable to run on Mac JPype Demo_Javax.swing Import

### DISCUSSION THREAD

**Discussion by vsquared** [jpype-project/jpype#1191] [General]
I am unable to run this demo which came from 'JPype User Guide' on a mac (Sonoma 14.4.1, Intel chip).  Is there a known fix?
```python
import jpype
import jpype.imports

jpype.startJVM()
import java
import javax
from javax.swing import *

def createAndShowGUI():
    frame = JFrame("HelloWorldSwing")
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    label = JLabel("Hello World")
    frame.getContentPane().add(label)
    frame.pack()
    frame.setVisible(True)

# Start an event loop thread to handling gui events
@jpype.JImplements(java.lang.Runnable)
class Launch:
    @jpype.JOverride
    def run(self):
        createAndShowGUI()
javax.swing.SwingUtilities.invokeLater(Launch())
```
**Addendum:**
I'm running this code in a Thonny editor with the following JPype installed:
<img width="172" alt="jpype" src="https://github.com/jpype-project/jpype/assets/2294268/041c1230-fd3d-4987-a225-3ecce81529d6">

The error analysis by Thonny indicates that JFrame and JLabel are not defined.  Thanks for any help or insight.

**Comment by vsquared** [comment]
If we don't use the SwingUtilities thread and do something like the following it appears that the imports are being used, but still there is no frame displayed.  Are we missing a renderer or event loop?  I tried using asyncio but that made no difference.
```python
import jpype
import jpype.imports

jpype.startJVM()

import java
import javax

from javax.swing import *

def buildWnd():
    print("======= buildWnd begin =========")
    frame = JFrame("Hello Swing")
    frame.setBounds(100, 100, 400, 400)
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    frame.setVisible(True)
    print("frame =",frame)
    print("======= buildWnd end ===========")
    
if __name__ == '__main__':
    buildWnd()

```
output:
<img width="1060" alt="return" src="https://github.com/jpype-project/jpype/assets/2294268/8cb33b4e-a5d2-4041-9ecd-fa13b1244bb8">

**Comment by Thrameos** [comment]
All GUIs require a service thread.   That is something that takes events and routes them to a common place to be handled.   Your issue here is not JPype, but rather how to use swing.   You either need to tell the main thread to be adding as the servicing thread, or create a new Java thread and have it run the GUI servicing loop.     My recommendation would be test what you are trying to do in Java first because until you have properly set up the GUI servicing in Java, how to make it work through Python is going to be a challenge.

**Reply by vsquared** [reply]
To avoid confusion for other mac users who try to run this demo in the future, should there be a disclaimer in the Documentation that it does not run on a mac?  I saw another post from three or four years ago regarding the same issue, and I don't believe that was ever satisfactorily resolved.

**Reply by Thrameos** [reply]
Mac is a separate problem from Swing.   JPype does run on Mac but requires the AppHelper loop rather than the Swing loop.   The problem there is that only the main thread can be used to serve AppHelper.   Thus you have to create a function that you want to run, then use the  "setupGuiEnvironment" call which will launch your desired function in another thread.    This solution works and is in the documentation, but is exceptionally painful to use.   YMMV as the pattern for mac is so different from all others.

So for all operating systems except mac you can use any thread you want service the GUI loop.   For Mac only you must use the main python thread or it will dead lock.   It you want code that works on any then you must follow the mac pattern (which is exceptionally annoying because it makes interactive execution impossible.) 

I was trying to push through a JPype executable version that solves that issue by starting main thread with Java and Python as spawned thread.   Unfortunately as I have no access to an interactive node on a mac I can't do any debugging and no one has reported back the modifications required to work on Mac.  And all I can do is tell you what others have reported to me regarding the required threading.   Please note that none of these are JPype issues but rather interactions between the JVM and system GUI.   We are just an incent third party that is fighting limitations being imposed by larger kits.

As I said previously, you first need to understand how GUI servicing works.   Setting up windows and then exiting before the windows have ever been serviced does nothing.   Then once you know how to deal with servicing, then dealing with system that require specific threads to service will make sense.  We unfortunately do not have the resources to teach everyone how to properly operate a GUI as it is a much larger subject than the Python to Java bridge.

**Comment by vsquared** [comment]
Thanks for the tip on AppHelper.  The following now runs on my Mac (Sonoma 14.4.1 Intel) using a Thonny editor:
```python
import jpype
import jpype.imports

jpype.startJVM()

import java
import javax
from javax.swing import *
from PyObjCTools import AppHelper

def buildWnd():
    print("buildWnd called.")
    print("frame in buildWnd =",frame)
    btn = javax.swing.JButton("Button")
    btn.setBounds(30,60,100,24)
    frame.add(btn)
    frame.setVisible(True)
    print("btn =",btn)
    print("EDT =",javax.swing.SwingUtilities.isEventDispatchThread())
    print("Done in buildWnd.")
    
def main():
    global frame

    frame = javax.swing.JFrame("Swing EDT")
    frame.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE)
    frame.setBounds(100,100,400,400)
    frame.setLayout(None)

    javax.swing.SwingUtilities.invokeLater(buildWnd)
    AppHelper.runEventLoop()

main()
```
**Output:**

<img width="405" alt="helper" src="https://github.com/jpype-project/jpype/assets/2294268/ffbc0dfb-f132-41cf-8bd1-8262aa188a24">

**Console Output:**
```
buildWnd called.
frame in buildWnd = 2024-06-05 20:53:27.973 Python[39738:1728535] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.
javax.swing.JFrame[frame0,100,100,400x400,invalid,hidden,layout=java.awt.BorderLayout,title=Swing EDT,resizable,normal,defaultCloseOperation=EXIT_ON_CLOSE,rootPane=javax.swing.JRootPane[,0,0,0x0,invalid,layout=javax.swing.JRootPane$RootLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=16777673,maximumSize=,minimumSize=,preferredSize=],rootPaneCheckingEnabled=true]
btn = javax.swing.JButton[,30,60,100x24,alignmentX=0.0,alignmentY=0.5,border=com.apple.laf.AquaButtonBorder$Dynamic@164b472f,flags=288,maximumSize=,minimumSize=,preferredSize=,defaultIcon=,disabledIcon=,disabledSelectedIcon=,margin=javax.swing.plaf.InsetsUIResource[top=0,left=2,bottom=0,right=2],paintBorder=true,paintFocus=true,pressedIcon=,rolloverEnabled=false,rolloverIcon=,rolloverSelectedIcon=,selectedIcon=,text=Button,defaultCapable=true]
EDT = True
Done in buildWnd.
```
**Addendum:**
This also works on a Mac:
```python
import jpype
import jpype.imports

jpype.startJVM()
import java
import javax
from javax.swing import *
from Cocoa import NSApp

def createAndShowGUI():
    frame = JFrame("HelloWorldSwing")
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    frame.setBounds(100, 100, 400, 400)
    label = JLabel("Hello World")
    frame.getContentPane().add(label)
    frame.pack()
    frame.setVisible(True)

javax.swing.SwingUtilities.invokeLater(createAndShowGUI)
NSApp.run()
```
